pkg/asset/cluster/cluster.go

```diff
@@ -0,0 +1,101 @@
+package cluster
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/terraform"
+	"github.com/openshift/installer/pkg/types/config"
+)
+
+// Cluster uses the terraform executable to launch a cluster
+// with the given terraform tfvar and generated templates.
+type Cluster struct {
+	// The root directory of the generated assets.
+	rootDir string
+	tfvars  asset.Asset
+
+	// TODO(yifan): Add kubeconfig-admin as a dependency
+	// so we can use it to determine when the cluster is
+	// fully up.
+}
+
+var _ asset.Asset = (*Cluster)(nil)
+
+// Name returns the human-friendly name of the asset.
+func (c *Cluster) Name() string {
+	return "Cluster"
+}
+
+// Dependencies returns the direct dependency for launching
+// the cluster.
+func (c *Cluster) Dependencies() []asset.Asset {
+	return []asset.Asset{c.tfvars}
+}
+
+// Generate launches the cluster and generates the terraform state file on disk.
+func (c *Cluster) Generate(parents map[asset.Asset]*asset.State) (*asset.State, error) {
+	state, ok := parents[c.tfvars]
+	if !ok {
+		return nil, fmt.Errorf("failed to get terraform.tfvar state in the parent asset states")
+	}
+
+	var tfvars config.Cluster
+	if err := json.Unmarshal(state.Contents[0].Data, &tfvars); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal terraform tfvars file: %v", err)
+	}
+
+	steps = []string{
+		terraform.InfraStep,
+		terraform.BootstrapStep,
+	}
+
+	dir, err := baseLocation()
+	if err != nil {
+		return nil, err
+	}
+
+	var result asset.State
+
+	for _, step := range steps {
+		templateDir, err := terraform.FindStepTemplates(dir, step, tfvars.Platform)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := terraform.Init(c.rootDir, templateDir); err != nil {
```

Comment:
nit: `terraform.Apply` might do the init internally?

Reply:
sure.

Reply:
Hmm, maybe just leave as is for now?